### PR TITLE
changes for leasing script

### PR DESCRIPTION
**Pull request by ravisantoshgudimetla**
Sorry for trouble. I think I forgot to rebase my fork earlier. I did that now and it should pass all the tests. I closing the earlier pull request #587

### CODE REVIEW

**Comment by matsuuran** [comment]
@ravisantoshgudimetla 
is this just waiting on folks to review or are we waiting for another milestone to integrate?

**Comment by ravisantoshgudimetla** [comment]
@matsuuran 
I am waiting for a review.

File: scripts/dynamic_free_pool.py [hunk 1]
```diff
@@ -0,0 +1,183 @@
+#! /bin/python
+import urlparse
+import requests
+from sets import Set
+import json
+import time
+import sys
+from requests.auth import HTTPBasicAuth
+import ConfigParser, os
+#Release nodes : Takes node list as arguments and puts them back to free pool
```
Comment (knikolla) [inline scripts/dynamic_free_pool.py]: Can you please change all these comments into dosctrings?

File: scripts/status_file.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+sun-12 free_pool 0
```
Comment (knikolla) [inline scripts/status_file.txt]: Why is this file here?

File: scripts/dynamic_free_pool.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	free_node_list =  urlparse.urljoin(haas_url,'free_nodes')
+	free_node_list = requests.get(free_node_list, auth = (haas_user_name, haas_password))
+	return free_node_list
+
+
+#For checking purpose
+if __name__ == "__main__":
+	try:
+		config = ConfigParser.SafeConfigParser()
+		config.read("/home/ravi/leasing.cfg")
```
Comment (knikolla) [inline scripts/dynamic_free_pool.py]: this should come from a command line argument.

Comment (ravisantoshgudimetla) [inline scripts/dynamic_free_pool.py]: Actually this should be moved to /etc/, so that we can read from that location instead of having it in some user's home location. But as of now, since I am reading from my account, I did it that way. Also, we need to run it from haas or some other account.

File: scripts/status_file.txt (hunk 2 above)
Comment (ravisantoshgudimetla) [inline scripts/status_file.txt]: Added this as a sample file. I gave the format in documentation.

**Comment by ravisantoshgudimetla** [comment]
@knikolla : Made the changes you suggested.

**Comment by matsuuran** [comment]
@ravisantoshgudimetla I assume that this is waiting for review again

**Comment by matsuuran** [comment]
@knikolla @gsilvis @henn @SahilTikale @zenhack can anyone review this?

**Comment by shwsun** [comment]
@ravisantoshgudimetla cam you spend a little time to fix those pep8 errors?

File: scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
+# Leasing Scripts:
+
+
+Dynamic Free Pool:
+
+This script is used for reclaiming the non persistent nodes from projects in HaaS.
+
+Limitations:
+
+As of now, scripts location etc. are placed in my home folder. I want to change them to a service account(user which is not on my name). Also, this is more or less like a ticketing system which runs from a cron and executes the current script. So, there are chances of drifting if node gets just registered before
```
Comment (zenhack) [inline scripts/README.md]: Hard-wrap lines in the docs to 72-79 chars (we haven't agreed upon an exact value for the docs, 79 is the rule for source code). This can get annoying to deal with in a text editor.

This goes for everything; this is just the most obvious case.

File: scripts/dynamic_free_pool.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+			and project_in_haas != None:
+			update_project_in_status_file(statusfile, node, project_in_haas,\
+							project_in_file)
+				
+		#else:
+			#Need to write a function which actually brings down 
+			#the nodes if they are in free pool.
+			#Power save mode on :)	
+
+def compare_projects(project_in_file,project_in_haas):
```
Comment (zenhack) [inline scripts/dynamic_free_pool.py]: This function is literally just `==`; why even have it?

**Comment by zenhack** [comment]
@ravisantoshgudimetla, can you provide background? Questions:

1. What is this script for?
2. Why is it being added to the mainline HaaS repo? My first impression is that this is the sort of thing that belongs in the `examples` directory if it belongs in the repo at all.

I want to understand why we're even doing this before I pick through this in more detail. In terms of quality, this is nowhere near acceptable for the mainline repo -- hard-coded paths to your home directory, numerous pep8 violations, and some just silly bits of code (I called out one example). But I'm not going to ask you to go fix all the small things until I understand the big picture.

**Comment by ravisantoshgudimetla** [comment]
@zenhack : 
1. This is a leasing script which claims the nodes from a  HIL project and puts them back to HIL free pool after some time(mentioned in README.md).
2. Initially, this script was thought to be at a layer above HIL and use HIL API for getting the information related to nodes but after some more discussion, this was asked to be put in HIL main repo.

Coming to quality, I agree that I have to fix PEP8 violations and hard coding is something that I wanted to avoid but the only other options is to pass it from command line arguments and since this was kind of script that's going to run on top of HIL, we didn't bother much. I am open to other suggestions too. This was something that I mentioned in README.md too. I was using functional decomposition to avoid code repetition for the silly bits of code that you mentioned.  I prefer smaller functions which I think improves readability. Again, I don't have reservations on this, if you want me to remove this, I can. 

Please let me know, if you need additional information.

**Comment by shwsun** [comment]
I think we should discuss whether or not it is some features we need first. 

//cc: @henn @SahilTikale @gsilvis

File: scripts/dynamic_free_pool.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+                        status_file.write(line)
+	time.sleep(5)
+	response = release_node_from_project(haas_url, haas_username, haas_password,\
+						node, project)
+	return response	
+		
+def release_node_from_project(haas_url, haas_username, haas_password, node, project):
+	"""Release a node from a project."""
+	haas_url = haas_url + 'project/' + project + '/detach_node'
+	body = {
```
Comment (henn) [inline scripts/dynamic_free_pool.py]: This can be one line

**Comment by henn** [comment]
@zenhack, @shwsun - we asked @ravisantoshgudimetla for this script so that certain nodes could be made available on a temporary basis in order to prevent hoarding/accidentally not releasing nodes (which is what currently is happening).

I think the HIL repo is a good place for it for now, as it a) provides a useful feature and b) gives an example of a script that uses the REST API.
In terms of which directory, I agree that maybe `examples/` would be better.

**Comment by shwsun** [comment]
It makes sense to me. @henn @ravisantoshgudimetla  

However, if we will have this feature, 
1) we have to follow pep8 regulations, i.e. code indention has to be 4.
2) Some code clean process will be needed, i.e. dead code should be removed and adding some comment would be good for us to do code review and testing. 
3) Some more documentation might be needed?

**Comment by henn** [comment]
@ravisantoshgudimetla - could you move this to `examples/`?

You can do that using `git mv <filename> ../examples/`

**Comment by zenhack** [comment]
I'm sold on something filling this repo being in the repo. I do feel strongly that it go in `examples/`.

As has been pointed out, there's a lot of cleanup needed. Please do the things that have already been pointed out, (formatting and such) and these additional general changes:

 * Convert calls functions like `compare_projects` to just usages of the corresponding operator (`==`, `<`) etc, and delete the definitions, 
* Delete commented out code; in general, this kind of thing is fine for debugging, but don't commit it to the repo.

Once you've done that I'll give it a more careful review.

File: scripts/dynamic_free_pool.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+	return free_node_list
+
+
+if __name__ == "__main__":
+	"""Main function for calling rest of functions"""
+	try:
+		config = ConfigParser.SafeConfigParser()
+		#Leaving it for as of now and has to be replaced by another 
+		#location which HIL has access to.
+		config.read("/home/ravi/leasing.cfg")
```
Comment (knikolla) [inline scripts/dynamic_free_pool.py]: maybe `/etc/hil_leasing.cfg` ?

File: scripts/dynamic_free_pool.py (hunk 5 above)
Comment (knikolla) [inline scripts/dynamic_free_pool.py]: ++

**Comment by naved001** [comment]
Hi @ravisantoshgudimetla 
Have you had time to look into the changes requested for this pr? Thanks!

**Comment by shwsun** [comment]
@ravisantoshgudimetla any updates? HIL team are considering changing from `HaaS` to `HIL` once we finished all the PR .

**Comment by ravisantoshgudimetla** [comment]
@shwsun , @naved001 , I am almost done with the changes, I will put the file in examples directory so that people can review again by end of this week.

**Comment by naved001** [comment]
@ravisantoshgudimetla any updates on this pr?

**Comment by ravisantoshgudimetla** [comment]
@naved001 , I am going to close this and open it again as I don't want to block the HIL name change. I am currently not able to work on this, hopefully in next two weeks, I will open new one.
